### PR TITLE
feat(openapi): create api specification

### LINKED ISSUE #3: create api specification
We should create an API specification, preferably in an open format like [OpenAPI](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md)

### PR DESCRIPTION
**Pull request by KnutZuidema**
closes #3

### CODE REVIEW

**Review by jnnkls — DISMISSED** [review]
Sounds good!

**Comment by KnutZuidema** [comment]
@JaniDJ any thoughts on the endpoints I specified? Do you think we will need anything else? I would like this to be basically complete so we can work independently on the front- and backend without adjusting things later.

File: openapi.yaml [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+      summary: requesting all groups
+      operationId: getGroups
+      responses:
+        200:
+          description: successfully returned all groups
+          content:
+            application/json:
+              schema:
+                items:
+                  $ref: '#/components/schemas/Group'
```
Comment (frzifus) [inline openapi.yaml:21]:
```suggestion
                  $ref: '#/components/schemas/group'
```

File: openapi.yaml [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+      operationId: createGroup
+      requestBody:
+        $ref: '#/components/requestBodies/group'
+      responses:
+        201:
+          description: successfully created group
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Group'
```
Comment (frzifus) [inline openapi.yaml:76]:
```suggestion
                $ref: '#/components/schemas/group'
```

File: openapi.yaml [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+      operationId: getPerson
+      parameters:
+        - $ref: '#/components/parameters/id'
+      responses:
+        200:
+          description: returns a person
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Person'
```
Comment (frzifus) [inline openapi.yaml:91]:
```suggestion
                $ref: '#/components/schemas/person'
```

File: openapi.yaml [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        $ref: '#/components/requestBodies/person'
+      parameters:
+        - $ref: '#/components/parameters/id'
+      responses:
+        200:
+          description: successfully updated person
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/Person'
```
Comment (frzifus) [inline openapi.yaml:107]:
```suggestion
                $ref: '#/components/schemas/person'
```

File: openapi.yaml [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+      requestBody:
+        $ref: '#/components/requestBodies/person'
+      responses:
+        201:
+          description: successfully created person
+          content:
+            application/json:
+              schema:
+                allOf:
+                  - $ref: '#/components/schemas/Person'
```
Comment (frzifus) [inline openapi.yaml]:
```suggestion
                  - $ref: '#/components/schemas/person'
```

File: openapi.yaml [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+        application/json:
+          schema:
+            type: object
+            properties:
+              surname:
+                type: string
+              name:
+                type: string
+              major:
+                $ref: '#/components/schemas/Major'
```
Comment (frzifus) [inline openapi.yaml]:
```suggestion
                $ref: '#/components/schemas/major'
```

File: openapi.yaml [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+              gender:
+                $ref: '#/components/schemas/Gender'
```
Comment (frzifus) [inline openapi.yaml]:
```suggestion
                $ref: '#/components/schemas/gender'
```

File: openapi.yaml [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+      type: object
+    Group:
+      properties:
+        id:
+          $ref: '#/components/schemas/uuid'
+        name:
+          type: string
+        members:
+          items:
+            $ref: '#/components/schemas/Person'
```
Comment (frzifus) [inline openapi.yaml:203]:
```suggestion
            $ref: '#/components/schemas/person'
```

File: openapi.yaml [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+        groupId:
+          $ref: '#/components/schemas/uuid'
+        surname:
+          type: string
+        name:
+          type: string
+        major:
+          $ref: '#/components/schemas/Major'
+        gender:
+          $ref: '#/components/schemas/Gender'
```
Comment (frzifus) [inline openapi.yaml:186]:
```suggestion
          $ref: '#/components/schemas/gender'
```

File: openapi.yaml [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+        id:
+          $ref: '#/components/schemas/uuid'
+        groupId:
+          $ref: '#/components/schemas/uuid'
+        surname:
+          type: string
+        name:
+          type: string
+        major:
+          $ref: '#/components/schemas/Major'
```
Comment (frzifus) [inline openapi.yaml:184]:
```suggestion
          $ref: '#/components/schemas/major'
```

Comment (KnutZuidema) [reply openapi.yaml:184]: The openAPI generator for the golang server doesn't like it when the schema names start with a lower-case letter.

**Comment by jnnkls** [comment]
On which site should the group assignment take place? If we want to assign groups on the server site I would recommend an "/groups/:groupId/users/:userId/add" endpoint to directly modify group/user relationship. And we need a place to set some kind of permissions like "student" and "tutor" for the different pages.

**Comment by KnutZuidema** [comment]
Group assignment should be server-side, IMO. Otherwise people could just assign themselves to any group if they know the ID. Member can not be removed from groups though, so I don't think we need a way to modify group-membership.
You are right, there should be a tutor assigned to each group, I haven't thought of that. Good point!
However, I believe tutor assignment should be entirely server-side as well, so I think we don't need another endpoint for that.
Happy to hear your thoughts on this.

**Comment by jnnkls** [comment]
Right so, let's think about this for a minute. A user opens the site and registers with surname, name, major, gender and age. The frontend calls the create user endpoint and gets a new user with a specific group. The group endpoint should return the corresponding tutors and an image then, right? Should there be an admin page for deleting fake users? Do tutors have some kind of login? The application is supposed to have some kind of presence check for each group.

**Comment by KnutZuidema** [comment]
I am not sure whether we should be concerned about people falsely registering users, since we have basically no way to check who is valid and who isn't, unless we could get a list of matriculation numbers, which we probably won't.
I don't think there is a login required for tutors or students. I believe the easiest way to do this is to just make it all public and possibly provide a way to search for a name, in which case we would need an endpoint for searching.
I don't think I understand what you mean by presence check.

**Comment by jnnkls** [comment]
> I don't think I understand what you mean by presence check.

I refer to the initial tasks description (on telegram) 

> Die tutoren verfügen über eine extra seite wo sie die ihenn zugeteilten erstsemester
einsehen und zur erleichterung der anwesenheitsüberprüfung makieren können.

**Comment by jnnkls** [comment]
> I am not sure whether we should be concerned about people falsely registering users, since we have basically no way to check who is valid and who isn't, unless we could get a list of matriculation numbers, which we probably won't.
I don't think there is a login required for tutors or students. I believe the easiest way to do this is to just 

I wasn't sure about how bullet prove this app should be 😉

**Comment by KnutZuidema** [comment]
True, it says that in the requirements, however I think a simple page showing all members of a group would be sufficient to check if all members are present.
I don't see an advantage in providing an additional page for just that purpose.
@frzifus what do you think?

**Comment by KnutZuidema** [comment]
After some consideration, I think that an endpoint for creating a group is not needed either. If group assignment is server-side, there is no sense in keeping group creation client-side. We will know the amount of groups needed and the tutor who will be assigned beforehand, so those could be manually created. We could allow group creation and assignment of tutors, as well as editing existing groups etc. via the API only if we had some kind of authorization, which right now we wouldn't need. I am not opposed to implementing an auth flow, so if you guys are fine with it I'll add it to the spec

**Comment by jnnkls** [comment]
If we assume that we know the number of groups and corresponding tutors a setup without auth flow would be fine with me.

**Comment by KnutZuidema** [comment]
I contacted the data processing center (DVZ) to see if we could use the OAuth service provided for FH members. That way we could actually check whether an entry is a valid student. I will update you all here once I know more.

**Comment by frzifus** [comment]
The number of tutors is known in advance. So we can assign the students to the tutors based on our requirements. I think a greedy way is suitable for doin this. 
Regarding the list we get the information from the group endpoint. The presence check list can be completely local.

any updates @KnutZuidema ?

**Comment by KnutZuidema** [comment]
Just as an update:
I opened a ticket for the request, however they do not provide an OAuth service yet. They are open to discussing the possibility though. However, I found out that apparently there is already a [Shibboleth](https://www.shibboleth.net/) service in place which would also provide what we need. I will call the DVZ to discuss those possibilities and will update you as soon as I have new information.

**Comment by KnutZuidema** [comment]
Alright, so, I just got off a phone call with the DVZ and they are generally open to letting us use the Shibboleth service. If we do want to use it, I will let them know and we can set up a meeting to discuss how using that service will look like. Does anyone have any objections on doing so?
@frzifus @JaniDJ

**Comment by KnutZuidema** [comment]
@JaniDJ @frzifus do you want to proceed with this course of action? I would wait until after the exams period to actually meet with the DVZ, but if we agree on doing this, I could already schedule that meeting.

**Comment by frzifus** [comment]
Let's set up a meeting between 19-21.07.2019.

**Review by frzifus — APPROVED** [review]

**Pull request merged**